### PR TITLE
bpt rate

### PR DESCRIPTION
**Pull request by tonkers-kuma**

### CODE REVIEW

**Review by newmickymousse — COMMENTED** [review]
Leave some comments to review

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -365,12 +356,17 @@ contract Strategy is BaseStrategy {
...
-        balancerVault.exitPool(balancerPoolId, address(this), address(this), request);
+    function sellBpt(uint256 _amountBpts) external onlyVaultManagers {
+        _sellBpt(_amountBpts);
+    }
+
+    // sell bpt for want at current bpt rate
+    function _sellBpt(uint256 _amountBpts) internal {
+        if (_amountBpts > 0) {
+            bytes memory userData = abi.encode(IBalancerVault.ExitKind.EXACT_BPT_IN_FOR_ONE_TOKEN_OUT, Math.min(_amountBpts, balanceOfBpt()), tokenIndex);
+            IBalancerVault.ExitPoolRequest memory request = IBalancerVault.ExitPoolRequest(assets, minAmountsOut, userData, false);
```
[the remark this answers is not in the document]
Comment (tonkers-kuma) [reply contracts/Strategy.sol:367]: it's always 0. Do you think we should enforce a min?

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -335,21 +322,25 @@ contract Strategy is BaseStrategy {
...
-                totalWantPooled += tokenPooled;
-            }
-        }
-        return totalWantPooled;
+        return bptsToTokens(balanceOfBpt());
+    }
+
+    /// use bpt rate to estimate equivalent amount of want.
+    function bptsToTokens(uint _amountBpt) public view returns (uint _amount){
+        uint unscaled = _amountBpt * bpt.getRate() / 1e18;
```
[the remark this answers is not in the document]
Comment (tonkers-kuma) [reply contracts/Strategy.sol:330]: the rate is calculated from invariant/totalSupply.
the invariant assumes all assets in the pool are the same price. So pool manipulation of the pool balances doesn't change the invariant. According to fish, it's not manipulatable

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -170,18 +171,23 @@ contract Strategy is BaseStrategy {
 
         // 2 forms of profit. Incentivized rewards (BAL+other) and pool fees (want)
         collectTradingFees();
-        sellRewards();
+        _sellRewards();
```
[the remark this answers is not in the document]
Comment (tonkers-kuma) [reply contracts/Strategy.sol:174]: I can add some toggles for this and the require check the FP mentioned

File: contracts/Strategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (newmickymousse) [reply contracts/Strategy.sol:367]: Well there are two things: 1) if we are using it as 0s, we could delete the variable at a contract level. 2) I think this is the way to check slippage, right? It's what we wait the pool to give us when we want to be out. So, it should solve part of our issues with slippage out.

File: contracts/Strategy.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (newmickymousse) [reply contracts/Strategy.sol:330]: Okay, great!

File: contracts/Strategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (tonkers-kuma) [reply contracts/Strategy.sol:380]: hm you're right, now with only 1 type of exit, we could replace the `enforceSlippageOut` with this built in one instead

File: tests/test_operation.py [hunk 4]
```diff
@@ -209,3 +215,79 @@ def test_rewards(
...
+        vault.withdraw(vault.balanceOf(user) / 2, user, 10000, {"from": user})
+    old_slippage = strategy.maxSlippageOut()
+
+    # loosen the slippage check to let the lossy withdraw go through
+    strategy.setParams(10000, 10000, strategy.maxSingleDeposit(), strategy.minDepositPeriod(), {'from': gov})
+    vault.withdraw(vault.balanceOf(user) / 2, user, 10000, {"from": user})
+    print(f'user balance: {token.balanceOf(user)}')
+    print(f'user lost: {amount / 2 - token.balanceOf(user)}')
+    util.stateOfStrat("after lossy withdraw", strategy, token)
+
```
Comment (newmickymousse) [inline tests/test_operation.py:283]: Hey there seems to be a merge problem here!

**Review by newmickymousse — COMMENTED** [review]
There is one thing to solve, otherwise LGTM

**Comment by wavey0x** [comment]
LGTM

**Comment by newmickymousse** [comment]
LGTM

**Comment by fp-crypto** [comment]
LGTM

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -190,59 +206,41 @@ contract Strategy is BaseStrategy {
         }
 
         uint256 pooledBefore = balanceOfPooled();
```
Comment (TomAFrench) [inline contracts/Strategy.sol:208]:
```suggestion
```
We don't use this variable anymore so no need to fetch it.

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -190,59 +206,41 @@ contract Strategy is BaseStrategy {
...
-        uint256[] memory maxAmountsIn = new uint256[](numTokens);
         uint256 amountIn = Math.min(maxSingleDeposit, balanceOfWant());
+        uint256 expectedBptOut = tokensToBpts(amountIn).mul(basisOne.sub(maxSlippageIn)).div(basisOne);
+        uint256[] memory maxAmountsIn = new uint256[](numTokens);
         maxAmountsIn[tokenIndex] = amountIn;
 
         if (amountIn > 0) {
             uint256[] memory amountsIn = new uint256[](numTokens);
             amountsIn[tokenIndex] = amountIn;
```
Comment (TomAFrench) [inline contracts/Strategy.sol:216]: We can reuse maxAmountsIn rather than recreating it here.

**Review by rabmarut — COMMENTED** [review]
Hey everyone, I'm pleading ignorance on certain implementation details of a yearn strategy, but I can say that all the Balancer-specific changes introduced in this PR are sound. I just want to highlight a couple of assumptions to make sure we're on the same page:

- This strategy should only be used for Balancer pools employing the StableSwap invariant. At time of writing, those would be StablePools and MetaStablePools. The tests imply that this is the intention, but I just want to make sure. This strategy, as written, could not be generalized to other types of Balancer pools like WeightedPools because it makes price correlation assumptions among the pool's underlying tokens and its BPT.
- The `getRate` function provides only an approximation of the price of one BPT in terms of any of the underlying tokens, assuming that all underlying prices are equivalent. This will never perfectly correlate to the actual balances in a StablePool, but assuming none of the underlying are way off peg it should be close enough for this and many other use cases.

On an unrelated note, this PR seems to introduce unchecked arithmetic in a few places, but I would leave that up to the author whether it is worth tightening up.

**Comment by wavey0x** [comment]
Thank you @rabmarut for taking time to review.
Yes, we're on the same page as you on each point. 

We rely on`.getRate()` in the same way that we rely on Curve's `.get_virtual_price()`. Essentially a way to derive the underlying asset value to/from the LP with the important assumption that the assets are **_perfectly_** pegged (in reality never). 

It is effectively a benchmark we can use at any time to help calculate how much profit/loss to attribute to the strategy's position upon each depost or withdrawal from the pool. This includes withdrawals that occur naturally during harvests, and also less-naturally through user withdrawals (including flashloan attacks).

**Comment by wavey0x** [comment]
@rabmarut + @TomAFrench  can you please ping me on telegram or twitter? user: @wavey0x

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -170,79 +176,68 @@ contract Strategy is BaseStrategy {
...
         } else {
             _loss = _loss.sub(_profit);
+            _debtPayment += _profit;
             _profit = 0;
         }
+
+        if (toggles.checkDebtOutstanding) {
+            // final check to make sure accounting is correct
+            require(_debtOutstanding == _debtPayment.add(_loss));
+        }
```
Comment (fp-crypto) [inline contracts/Strategy.sol:200]: This is in the wrong method, it should be in `liquidatePosition`, not prepare return

**Comment by newmickymousse** [comment]
LGTM
